### PR TITLE
Add support for all input methods in config file

### PR DESCRIPTION
**Pull request by shilangyu**

### CODE REVIEW

File: example/contracts_generator.yaml [hunk 1]
```diff
@@ -1,5 +1,20 @@
...
-input: ../contractsgenerator/examples/project/aggregated/Combined/Combined.csproj
+# Possible input methods, specify exactly one (required)
+input:
+  # Array of paths pointing to .csproj contracts project files
+  # project:
+  #   - ../contractsgenerator/examples/project/aggregated/Combined/Combined.csproj
+  # Path to a single source file
+  # file: ExampleContracts/Auth.cs
+  # Multiple files selected by glob patterns
+  path:
```
Comment (jakubfijalkowski) [inline example/contracts_generator.yaml:9]: Try to favour `project` source - it will be slightly slower in generation, but will be easier to manage (if someone decides to split the project into smaller ones, e.g. `common` + `core`).

File: example/contracts_generator.yaml [hunk 2]
```diff
@@ -1,5 +1,20 @@
-# Path to contracts' .csproj project file (required)
-input: ../contractsgenerator/examples/project/aggregated/Combined/Combined.csproj
+# Possible input methods, specify exactly one (required)
+input:
+  # Array of paths pointing to .csproj contracts project files
+  # project:
+  #   - ../contractsgenerator/examples/project/aggregated/Combined/Combined.csproj
+  # Path to a single source file
+  # file: ExampleContracts/Auth.cs
```
Comment (jakubfijalkowski) [inline example/contracts_generator.yaml:7]: I _think_ you won't need it, as this was created for testing mainly.

File: lib/src/contracts_generator_config.dart [hunk 3]
```diff
@@ -21,18 +21,74 @@ class ContractsGeneratorConfig {
 
     final config = mergeMaps<dynamic, dynamic>(defaultConfig, override);
 
+    late GeneratorScript script;
+
     if (config['input'] == null) {
       throw ArgumentError.notNull(
         '`input` is a required field and cannot be null',
       );
+    } else {
```
Comment (jakubfijalkowski) [inline lib/src/contracts_generator_config.dart:30]: A slight refactoring would be beneficial here, e.g. extracting a function for each case & changing the validation to something more direct.

File: example/contracts_generator.yaml (hunk 2 above)
Comment (shilangyu) [reply example/contracts_generator.yaml:7]: By the way, isn't `generate.sh file -i some/file.cs` and `generate.sh path -i some/file.cs` equivalent? Perhaps the whole `file` subcommand can be removed to avoid confusion

Comment (jakubfijalkowski) [reply example/contracts_generator.yaml:7]: They are, although they use slightly different ways of finding things that is (was?) beneficial for preliminary testing. Will probably remove soon.

File: lib/src/contracts_generator_config.dart (hunk 3 above)
Comment (shilangyu) [reply lib/src/contracts_generator_config.dart:30]: > changing the validation to something more direct.

Could you elaborate?

Comment (jakubfijalkowski) [reply lib/src/contracts_generator_config.dart:30]: So, now you have validation & parsing intertwined. This is quite hard to follow and test thoroughly. It is much easier to write (and maintain) validation as a part of parsing, i.e. you have three methods that each considers only a single _way_ of configuration. Doing _magic_ on a map, with `length` check, is rather _magic_ and quite hard to follow and extend.

To visualise what I am talking about, see this pseudocode:
```dart
GeneratorScript configureProject(dynamic config) {
  if (config['input']['project'] != null) {
    return GeneratorScript.project(config['input']['project'].value);
  } else {
    return null;
  }
}

// others look alike

GeneratorScript configure(dynamic config) {
  final proj = configureProject(config);
  final file = configureFile(config);
  final path = configurePath(config);
  
  // This check should be handled by the argument parsing library, not you
  if (proj != null ^ file != null ^ path != null) {
    throw ArgumentException('invalid config');
  }

  return proj ?? file ?? path;
}

// And the `fromYaml`
factory ContractsGeneratorConfig.fromYaml(String yaml) {
    final defaultConfig = loadYaml(defaultYamlConfig) as Map;
    final override = loadYaml(yaml) as Map;

    final config = mergeMaps<dynamic, dynamic>(defaultConfig, override);
    final dynamic name = config['name'];
    final dynamic include = config['include'];
    final dynamic output = config['output'];
    final dynamic extra = config['extra'];

    if (name is! String) {
      throw ArgumentError('`name` field has to be a string');
    }
    if (include is! String) {
      throw ArgumentError('`include` field has to be a string');
    }
    if (output is! String) {
      throw ArgumentError('`output` field has to be a string');
    }
    if (extra is! String) {
      throw ArgumentError('`extra` field has to be a string');
    }

    return ContractsGeneratorConfig(
      input: configure(config),
      name: name,
      include: RegExp(include),
      output: Directory(output),
      extra: extra,
    );
}
```

**Review by jakubfijalkowski — APPROVED** [review]
